### ISSUE TITLE
[XLA:GPU] TopK custom kernel (__gpu$TopK for k <= 16) violates IEEE 754 TotalOrder and corrupts output when -NaN is present

### ISSUE THREAD

**Issue by apivovarov** [openxla/xla#48119]
When `topk` is executed on GPU with $k \le 16$ and $n \ge 1024$, `TopkSpecializer` routes the operation to the bespoke CUDA kernel `__gpu$TopK` (topk_kernel_cuda_common.cu.h). This kernel uses raw C++ hardware floating-point comparisons (`lhs.key > rhs.key`) rather than IEEE 754 `TOTALORDER`. 

If topk is lowered to sort + slice the result is correct (because it uses comparator direction=GT, type=TOTALORDER)

```
-- Actual Topk Result (__gpu$TopK) ---------------------------------------------
nan (Hex: 0xffc00000) at index 1023
nan (Hex: 0xffc00000) at index 959
nan (Hex: 0xffc00000) at index 895
nan (Hex: 0xffc00000) at index 831
nan (Hex: 0xffc00000) at index 767
nan (Hex: 0xffc00000) at index 703
nan (Hex: 0xffc00000) at index 639
nan (Hex: 0xffc00000) at index 575
nan (Hex: 0xffc00000) at index 511
nan (Hex: 0xffc00000) at index 447
nan (Hex: 0xffc00000) at index 383
nan (Hex: 0xffc00000) at index 319
nan (Hex: 0xffc00000) at index 255
nan (Hex: 0xffc00000) at index 191
nan (Hex: 0xffc00000) at index 127
nan (Hex: 0xffc00000) at index 63
```

```
-- Expected Topk Result (TOTALORDER) -------------------------------------------
nan (Hex: 0x7fe00001) at index 5
nan (Hex: 0x7fe00000) at index 19
nan (Hex: 0x7fc00000) at index 4
inf (Hex: 0x7f800000) at index 1
inf (Hex: 0x7f800000) at index 8
90.44499969482422 (Hex: 0x42b4e3d7) at index 18
0.012500000186264515 (Hex: 0x3c4ccccd) at index 20
9.99994610111476e-41 (Hex: 0x116c2) at index 21
1.0005271035279194e-42 (Hex: 0x2ca) at index 13
1.0005271035279194e-42 (Hex: 0x2ca) at index 23
0.0 (Hex: 0x0) at index 3
0.0 (Hex: 0x0) at index 16
-0.0 (Hex: 0x80000000) at index 6
-0.0 (Hex: 0x80000000) at index 12
-9.949219096706201e-44 (Hex: 0x80000047) at index 14
-9.949219096706201e-44 (Hex: 0x80000047) at index 24
-------------------------------------------------------------------
```

Test to reproduce the bug
```py
from absl.testing import absltest
import jax
from jax import export
import jax.numpy as jnp
import numpy as np
import struct

@jax.jit
def topk_fn(x):
  return jax.lax.top_k(x, k=16)

def generate_xla_total_order_topk(arr_f32, k):
  """Generates bit-for-bit accurate XLA Total Order Top-K using numpy."""
  # 1. View bits as signed 32-bit integers (matching XLA's bitcast to s32)
  s32_view = arr_f32.view(np.int32)

  # 2. TOTAL ORDER logic: if negative (s32 < 0), XOR with 0x7FFFFFFF, else keep as is.
  is_negative = s32_view < 0
  mask = np.where(is_negative, np.int32(0x7FFFFFFF), np.int32(0))
  comparable_s32 = s32_view ^ mask

  # 3. Sort descending stably. Cast to int64 before negating to prevent overflow
  # when negating -2147483648 (which would wrap around in int32).
  sort_keys = -comparable_s32.astype(np.int64)
  descending_indices = np.argsort(sort_keys, axis=-1, kind="stable")

  # 4. Slice the Top-K indices and gather the original floats
  topk_indices = descending_indices[..., :k]
  topk_values = np.take_along_axis(arr_f32, topk_indices, axis=-1)

  return topk_values, topk_indices

class TopKTest(absltest.TestCase):
  def test_topk_with_infinity_execution(self):
    print("--- Execution Test ------------------------------------------")
    device = jax.devices()[0]
    print(f"Running on device: {device}")
    k = 16
    N = 1023   # (sort(TOTALORDER) + slice) - GOOD
    N = 1024   # (custom-call __gpu$TopK) - BAD

    neg_qnan0_f32 = np.uint32(0xFFC00000).view(np.float32)
    neg_qnan1_f32 = np.uint32(0xFFE00000).view(np.float32)
    neg_qnan2_f32 = np.uint32(0xFFE00001).view(np.float32)

    pos_qnan0_f32 = np.uint32(0x7FC00000).view(np.float32)
    pos_qnan1_f32 = np.uint32(0x7FE00000).view(np.float32)
    pos_qnan2_f32 = np.uint32(0x7FE00001).view(np.float32)

    # Create an input tensor [Batch, Sequence]
    s = [
        #90.445,
        float("-inf"),
        #0.0125,
        float("+inf"),
        #1.0e-40,
        -1.0e-41,
        +0.0,
        pos_qnan0_f32,
        pos_qnan2_f32,
        -0.0,
        -90.445,
        float("+inf"),
        neg_qnan0_f32,
        neg_qnan2_f32,
        -90.445,
        -0.0,
        1.0e-42,
        -1.0e-43,
        neg_qnan1_f32,
        +0.0,
        float("-inf"),
        90.445,
        pos_qnan1_f32,
        0.0125,
        1.0e-40,
        -1.0e-41,
        1.0e-42,
        -1.0e-43,
    ]
    bot = s + [float("-nan")] * (N - len(s))
    batch_size = 1
    bot_batch = [bot] * batch_size

    x = jax.device_put(
        jnp.array(bot_batch, dtype=jnp.float32),
        device,
    )

    # Generate the dynamic golden truth using NumPy
    bot_batch_np = np.array(bot_batch, dtype=np.float32)
    expected_values, expected_indices = generate_xla_total_order_topk(
        bot_batch_np, k
    )

    # Cast expected floats to uint32 for strict bitwise assertion
    expected_values_u32 = expected_values.view(np.uint32)

    for i in range(200):
      # Run the function
      values2, indices2 = topk_fn(x)
      self.assertEqual(values2.shape, (batch_size, k))
      self.assertEqual(indices2.shape, (batch_size, k))
      # Compare indices directly
      np.testing.assert_array_equal(
          np.asarray(indices2),
          expected_indices,
          err_msg=f"Indices mismatched on iteration {i}!",
      )

      # Compare values bit-for-bit to catch payload corruptions
      actual_values_u32 = np.asarray(values2).view(np.uint32)
      np.testing.assert_array_equal(
          actual_values_u32,
          expected_values_u32,
          err_msg=f"Values mismatched on iteration {i} (bitwise check failed)!",
      )

    print("-- Actual Topk Result ---------------------------------------------")
    for i in range(k):
      val = values2[0, i]
      idx = indices2[0, i]
      # Pack the float into bytes, then unpack as a 32-bit unsigned int to see the hex
      hex_val = hex(struct.unpack("<I", struct.pack("<f", val))[0])
      print(f"{val} (Hex: {hex_val}) at index {idx}")

    print("-- Expected Topk Result -------------------------------------------")
    for i in range(k):
      val = expected_values[0, i]
      idx = expected_indices[0, i]
      # Pack the float into bytes, then unpack as a 32-bit unsigned int to see the hex
      hex_val = hex(struct.unpack("<I", struct.pack("<f", val))[0])
      print(f"{val} (Hex: {hex_val}) at index {idx}")
    print("-------------------------------------------------------------------")

if __name__ == "__main__":
  absltest.main()
